src/databricks/labs/ucx/hive_metastore/grants.py

```diff
@@ -434,12 +449,23 @@ def get_eligible_locations_principals(self) -> dict[str, dict]:
             logger.error(msg)
             raise ResourceDoesNotExist(msg) from None
 
-        for cluster_id, role_name in cluster_instance_profiles.items():
-            eligible_locations.update(self._get_external_locations(role_name, external_locations, permission_mappings))
+        for cluster_id, role_compute in cluster_instance_profiles.items():
+            eligible_locations = {}
+            eligible_locations.update(
+                self._get_external_locations(role_compute, external_locations, permission_mappings)
+            )
             if len(eligible_locations) == 0:
                 continue
-            cluster_locations[cluster_id] = eligible_locations
-        return cluster_locations
+            compute_locations.append(ComputeLocations(cluster_id, eligible_locations, "clusters"))
+        for warehouse_id, role_compute in warehouse_instance_profiles.items():
+            eligible_locations = {}
```

Comment:
Why create an empty dictionairy before the update?

Reply:
You could get 2 sp details from one cluster and 3 spn from second cluster, hence I am resetting the dict in the outer loop before the cluster loops to next cluster